src/lib/edit_script.rs

```diff
@@ -0,0 +1,297 @@
+                debug!("Edit script: UPD {} {:?} -> {} {:?}",
+                       store.from_arena[w].label,
+                       store.from_arena[w].value,
+                       store.to_arena[x].label,
+                       store.to_arena[x].value);
+                let mut upd = Update::new(w,
+                                          store.to_arena[x].value.clone(),
+                                          store.to_arena[x].label.clone());
+                upd.apply(&mut store.from_arena)?;
+                script.push(upd.clone());
+            }
+            // MOVE phase.
+            let y = store.to_arena[x].parent().unwrap();
+            let z = new_mappings.get_from(&y).unwrap();
+            if z != v {
+                let k = find_pos(store, x, &new_mappings, &to_in_order);
+                let mut mov = Move::new(w, z, k);
+                debug!("Edit script: MOV {} {:?} Parent: {} {:?}",
+                       store.from_arena[w].label,
+                       store.from_arena[w].value,
+                       store.from_arena[z].label,
+                       store.from_arena[z].value);
+                mov.apply(&mut store.from_arena)?;
+                script.push(mov);
+            }
+        }
+        align_children(store, w,
+                            x,
+                            &mut script,
+                            &new_mappings,
+                            &mut from_in_order,
+                            &mut to_in_order)?;
+    }
+    // Delete phase.
+    let mut actions = EditScript::new();
+    for w in root.post_order_traversal(&store.from_arena) {
+        if !new_mappings.contains_from(&w) {
+            debug!("Edit script: DEL {} {:?}",
+                   store.from_arena[w].label,
+                   store.from_arena[w].value);
+            let del = Delete::new(w);
+            script.push(del);
+            actions.push(del);
+        }
+    }
+    actions.apply(&mut store.from_arena)?;
+    // Add new mappings to the existing store.
+    debug!("Append new mappings to store.");
+    for (from, to) in new_mappings.from {
+        if !store.contains_from(&from) {
+            debug!("New mapping due to edit script: {} -> {}", from, to);
+            store.push(from, to, MappingType::EDIT);
+        }
+    }
+    Ok(script)
+}
+
+
+fn align_children<T: Clone + Debug + Eq + 'static>(store: &mut MappingStore<T>,
+                  w: NodeId,
+                  x: NodeId,
+                  script: &mut EditScript<T>,
+                  new_mappings: &TemporaryMappingStore,
+                  from_in_order: &mut HashSet<NodeId>,
+                  to_in_order: &mut HashSet<NodeId>)
+                  -> Result<(), ArenaError> {
+    debug!("align_children({}, {})", w, x);
+    for child in x.children(&store.to_arena) {
+        to_in_order.remove(&child);
+    }
+    let mut s1: Vec<NodeId> = vec![];
+    for child in w.children(&store.from_arena) {
```

Comment:
This `for` loop and the one below are identical except for the thing being iterated over. Maybe factor out into a closure?